packages/runtime/container-runtime/src/opLifecycle/outbox.ts

```diff
@@ -0,0 +1,191 @@
+                throw new GenericError(
+                    "BatchTooLarge",
+                    /* error */ undefined,
+                    {
+                        opSize: (message.contents?.length) ?? 0,
+                        count: this.attachFlowBatch.length,
+                        limit: this.attachFlowBatch.options.hardLimit,
+                    });
+            }
+        }
+    }
+
+    public flush() {
+        this.flushInternal(this.attachFlowBatch.popBatch());
+        this.flushInternal(this.mainBatch.popBatch());
+    }
+
+    private flushInternal(batch: IBatch) {
+        this.flushBatch(
+            this.addBatchMetadataToBatch(
+                this.processBatch(batch)));
+    }
+
+    private addBatchMetadataToBatch(batch: IBatch): IBatch {
+        if (batch.content.length > 1) {
+            batch.content[0].metadata = {
+                ...batch.content[0].metadata,
+                batch: true
+            };
+            batch.content[batch.content.length - 1].metadata = {
+                ...batch.content[batch.content.length - 1].metadata,
+                batch: false
+            };
+        }
+
+        return batch;
+    }
+
+    private processBatch(batch: IBatch): IBatch {
+        if (batch.content.length === 0 || this.options.compressionOptions === undefined) {
+            return batch;
+        }
+
+        let processedBatch = batch;
+        if (this.options.compressionOptions.minimumBatchSizeInBytes < batch.contentSizeInBytes) {
+            processedBatch = this.batchProcessors.compressor.processOutgoing(batch);
+        }
+
+        return processedBatch;
+    }
+
+    private flushBatch(batch: IBatch): void {
+        const length = batch.content.length;
+
+        if (length === 0) {
+            return;
+        }
+
+        let clientSequenceNumber: number = -1;
+
+        // Did we disconnect in the middle of turn-based batch?
+        // If so, do nothing, as pending state manager will resubmit it correctly on reconnect.
+        if (this.shouldSend()) {
+            clientSequenceNumber = this.sendBatch(batch.content);
+
+            // Convert from clientSequenceNumber of last message in the batch to clientSequenceNumber of first message.
+            clientSequenceNumber -= length - 1;
+            assert(clientSequenceNumber >= 0, 0x3d0 /* clientSequenceNumber can't be negative */);
+        }
+
+        this.persistPendingBatch(clientSequenceNumber, batch.content);
+    }
+
+    private sendBatch(batch: BatchMessage[]): number {
+        if (this.context.submitBatchFn === undefined) {
+            // Legacy path - supporting old loader versions. Can be removed only when LTS moves above
+            // version that has support for batches (submitBatchFn)
+            let clientSequenceNumber: number = -1;
+            for (const message of batch) {
+                // Legacy path doesn't support compressed payloads and will submit uncompressed payload anyways
```

Comment:
I'm not sure I follow this comment / code. If we are not supporting sending compressed payloads, we should not get here with compressed data, right? I.e. deleting property does not change much - content is still compressed

Reply:
This is copy pasted (and slightly adjusted) from the container runtime and deals with back-compat wrt to `submitBatchFn` vs `submitBatchFn`. I'll remove the comment here.